x-pack/plugin/repository-encrypted/src/main/java/org/elasticsearch/repositories/encrypted/EncryptedRepository.java

```diff
@@ -6,11 +6,359 @@
+        if (EncryptedRepositoryPlugin.getLicenseState().isEncryptedRepositoryAllowed()) {
+            if (consistentSettingsService.isConsistent(passwordSettingForThisRepo)) {
+                super.snapshotShard(store, mapperService, snapshotId, indexId, snapshotIndexCommit, snapshotStatus, writeShardGens, listener);
+            } else {
+                listener.onFailure(new RepositoryException(metadata.name(),
+                        "Password mismatch for repository [" + metadata.name() + "]. The local node's value of the " +
+                                "keystore secure setting [" + passwordSettingForThisRepo.getKey() + "] is different from the master's"));
+            }
+        } else {
+            listener.onFailure(LicenseUtils.newComplianceException(
+                    EncryptedRepositoryPlugin.REPOSITORY_TYPE_NAME + " snapshot repository"));
+        }
+    }
+
+    @Override
+    public void restoreShard(Store store, SnapshotId snapshotId, IndexId indexId, ShardId snapshotShardId,
+                             RecoveryState recoveryState, ActionListener<Void> listener) {
+        if (false == consistentSettingsService.isConsistent(passwordSettingForThisRepo)) {
+            // the repository has a different password on the local node compared to the master node
+            // even though restoring the shard will surely fail (because we know that, by now, the master's password
+            // must be correct, otherwise this method will not get called) we let it pass-through in order to avoid
+            // having to manipulate the {@code recoveryState} argument
+            logger.error("Password mismatch for repository [" + metadata.name() + "]. The local node's value of the " +
+                    "keystore secure setting [" + passwordSettingForThisRepo.getKey() + "] is different from the master's");
+        }
+        super.restoreShard(store, snapshotId, indexId, snapshotShardId, recoveryState, ActionListener.delegateResponse(listener,
+                (l, e) -> l.onFailure(new RepositoryException(metadata.name(), "Password mismatch for repository [" + metadata.name() +
+                        "]. The local node's value of the keystore secure setting [" +
+                        passwordSettingForThisRepo.getKey() + "] is different from the master's"))));
+    }
+
+    @Override
+    public void verify(String seed, DiscoveryNode localNode) {
+        if (consistentSettingsService.isConsistent(passwordSettingForThisRepo)) {
+            super.verify(seed, localNode);
+        } else {
+            // the repository has a different password on the local node compared to the master node
+            throw new RepositoryVerificationException(metadata.name(), "Repository password mismatch. The local node's [" + localNode +
+                    "] value of the keystore secure setting [" + passwordSettingForThisRepo.getKey() + "] is different from the master's");
+        }
+    }
```

Comment:
Verification fails nevertheless, overriding is not necessary, but this gets a friendlier error message:
```
{
  "error" : {
    "root_cause" : [
      {
        "type" : "repository_verification_exception",
        "reason" : "[test_enc] [[pN1b441YTV6fKug6AkF2Vg, 'org.elasticsearch.transport.RemoteTransportException: [ElasticMBP.local][127.0.0.1:9301][internal:admin/repository/verify]']]",
        "suppressed" : [
          {
            "type" : "repository_verification_exception",
            "reason" : "[test_enc] Repository password mismatch. The local node's [{ElasticMBP.local}{pN1b441YTV6fKug6AkF2Vg}{uHr4xFDrQzWtykhxV_VO9Q}{127.0.0.1}{127.0.0.1:9301}{dilm}{ml.machine_memory=17179869184, xpack.installed=true, ml.max_open_jobs=20}] value of the keystore secure setting [repository.encrypted.test_enc.password] is different from the master's"
          }
        ]
      }
    ],
    "type" : "repository_verification_exception",
    "reason" : "[test_enc] [[pN1b441YTV6fKug6AkF2Vg, 'org.elasticsearch.transport.RemoteTransportException: [ElasticMBP.local][127.0.0.1:9301][internal:admin/repository/verify]']]",
    "suppressed" : [
      {
        "type" : "repository_verification_exception",
        "reason" : "[test_enc] Repository password mismatch. The local node's [{ElasticMBP.local}{pN1b441YTV6fKug6AkF2Vg}{uHr4xFDrQzWtykhxV_VO9Q}{127.0.0.1}{127.0.0.1:9301}{dilm}{ml.machine_memory=17179869184, xpack.installed=true, ml.max_open_jobs=20}] value of the keystore secure setting [repository.encrypted.test_enc.password] is different from the master's"
      }
    ]
  },
  "status" : 500
}
```